File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    let total = 0;
    let l = 0, r = height.length - 1;
    let lmax = 0, rmax = height[r];

    while (l < r) {
        if (height[l] <= height[r]) {
            if (height[l] < lmax) {
                total += lmax - height[l]
            } else {
                lmax = height[l]
            }
            l++
        } else {
            if (height[r] < rmax) {
                total += rmax - height[r]
            } else {
                rmax = height[r]
            }
            r--
        }

    }
    return total;

};

// claude animation link
// https://claude.ai/artifacts/c1498e1c-187d-4681-9cc9-eab125d8bb9e

//claude chat link
//https://claude.ai/share/4b934750-027e-438a-9c58-2e6de0778cc3

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # total = 0
        # left = 0
        # right = len(height) - 1
        # lmax = 0
        # rmax = height[right]

        # while left < right:
        #     if height[left] <= height[right]:
        #         if height[left] < lmax:
        #             total += lmax - height[left]
        #         else:
        #             lmax = height[left]
        #         left +=1
        #     else:
        #         if height[right] < rmax:
        #             total += rmax - height[right]
        #         else:
        #             rmax = height[right]
        #         right -=1
        
        # return total
        l = 0
        r = len(height) - 1
        l_max = 0
        r_max = 0
        ans = 0

        while l < r:
            if height[l] < height[r]:
                l_max = max(l_max, height[l])
                ans += l_max - height[l]
                l += 1
            else:
                r_max = max(r_max, height[r])
                ans += r_max - height[r]
                r -= 1
        return ans

        
        